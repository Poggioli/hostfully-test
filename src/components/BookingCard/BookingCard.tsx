import { Booking } from "@/service/Booking";
import { FC } from "react";
import { Card, CardContent, CardDescription, CardFooter, CardHeader, CardTitle } from "@/components/ui/card";
import { format } from "date-fns";
import { DeleteBookingModal } from "../DeleteBookingModal";
import { AspectRatio } from "../ui/aspect-ratio";

type BookingCardProps = Booking;

const BookingCard: FC<BookingCardProps> = ({
  checkIn,
  checkOut,
  roomName,
  totalPrice,
  id,
  quantityGuests,
  photos
}) => {

  const bookingPeriod = `${format(checkIn, "PPP")} - ${format(checkOut, "PPP")}`

  return (
    <Card className="w-full">
      {photos?.length ? (
        <AspectRatio ratio={16 / 9} className="overflow-hidden mb-4 rounded-t-md">
          <img className="rounded-md object-cover w-full h-full" src={photos[0]} alt={`Photo of ${roomName}`} />
        </AspectRatio>
      ) : null}
      <CardHeader className={`flex flex-col pb-3 ${photos?.length ? 'pt-0' : ''}`}>
        <CardTitle>{roomName}</CardTitle>
        <CardDescription className="">{bookingPeriod}</CardDescription>
      </CardHeader>

      <CardContent className="flex flex-col gap-1 pb-4">
        <div className="flex gap-6 justify-between">
          <span className="underline">Total</span>
          <span>${totalPrice}</span>
        </div>
        <div className="flex gap-6 justify-between">
          <span className="underline">Guests</span>
          <span>{quantityGuests}</span>
        </div>
      </CardContent>

      <CardFooter className="flex justify-end">
        <DeleteBookingModal id={id} />
      </CardFooter>
    </Card>
  )
}

export { BookingCard };