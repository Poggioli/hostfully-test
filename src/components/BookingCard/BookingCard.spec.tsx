import { act, fireEvent, render } from "@testing-library/react";
import { BookingCard } from "./BookingCard";
import { QueryClient, QueryClientProvider } from "react-query";
import { ReactNode } from "react";

const mockOnEdit = jest.fn();
jest.mock("@/components/BookingDrawer", () => ({
  useBookingDrawer: () => ({
    onEdit: mockOnEdit,
  })
}));
const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
    mutations: {
      retry: false
    }
  },
})
const createWrapper = () => {
  return ({ children }: { children: ReactNode }) => (
    <QueryClientProvider client={queryClient}>
      {children}
    </QueryClientProvider>
  )
}

describe('BookingCard', () => {

  it(`GIVEN a BookingCard
      WHEN render
      THEN should render all props the right way`, () => {

    const { baseElement } = render(
      <BookingCard
        checkIn={new Date(2024, 0, 20)}
        checkOut={new Date(2024, 0, 24)}
        roomName="room name 1"
        roomId="room-id-1"
        totalPrice={5250}
        id="booking-id-1"
        quantityGuests={5}
        photos={[]}
      />, {
      wrapper: createWrapper()
    });

    expect(baseElement).toMatchSnapshot();
  });

  it(`GIVEN a BookingCard
      WHEN render
      THEN should render all props the right way`, () => {

    const { baseElement } = render(
      <BookingCard
        checkIn={new Date(2024, 0, 20)}
        checkOut={new Date(2024, 0, 24)}
        roomName="room name 1"
        roomId="room-id-1"
        totalPrice={5250}
        id="booking-id-1"
        quantityGuests={5}
        photos={['']}
      />, {
      wrapper: createWrapper()
    });

    expect(baseElement).toMatchSnapshot();
  });

  it(`GIVEN a BookingCard
      WHEN click to in button "Edit"
      THEN should call onEdit function from useBookingDrawer`, () => {

    const { getByRole } = render(
      <BookingCard
        checkIn={new Date(2024, 0, 20)}
        checkOut={new Date(2024, 0, 24)}
        roomName="room name 1"
        roomId="room-id-1"
        totalPrice={5250}
        id="booking-id-1"
        quantityGuests={5}
        photos={[]}
      />, {
      wrapper: createWrapper()
    }
    );

    const editButton = getByRole('button', {
      name: 'Edit'
    });

    act(() => fireEvent.click(editButton));

    expect(mockOnEdit).toHaveBeenCalledTimes(1);
    expect(mockOnEdit).toHaveBeenCalledWith("booking-id-1");
  });

});