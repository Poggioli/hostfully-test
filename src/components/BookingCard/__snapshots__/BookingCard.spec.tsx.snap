// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingCard GIVEN a BookingCard
      WHEN render
      THEN should render all props the right way 1`] = `
<body>
  <div>
    <div
      class="rounded-xl border bg-card text-card-foreground shadow w-full"
    >
      <div
        class="space-y-1.5 p-6 flex flex-col pb-3"
      >
        <h3
          class="font-semibold leading-none tracking-tight"
        >
          room name 1
        </h3>
        <p
          class="text-sm text-muted-foreground flex flex-row justify-between gap-x-6 flex-wrap"
        >
          <span>
            Jan 20, 2024 - Jan 24, 2024
          </span>
          <span>
            4 days
          </span>
        </p>
      </div>
      <div
        class="p-6 pt-0 flex flex-col gap-1 pb-4"
      >
        <div
          class="flex gap-6 justify-between"
        >
          <span
            class="underline"
          >
            Total
          </span>
          <span>
            $
            5250
          </span>
        </div>
        <div
          class="flex gap-6 justify-between"
        >
          <span
            class="underline"
          >
            Guests
          </span>
          <span>
            5
          </span>
        </div>
      </div>
      <div
        class="items-center p-6 pt-0 flex justify-between gap-4 flex-wrap"
      >
        <button
          aria-controls="radix-:r0:"
          aria-expanded="false"
          aria-haspopup="dialog"
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-destructive text-destructive-foreground shadow-sm hover:bg-destructive/90 h-9 px-4 py-2"
          data-state="closed"
          type="button"
        >
          <svg
            class="lucide lucide-trash mr-2 h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 6h18"
            />
            <path
              d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
            />
            <path
              d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
            />
          </svg>
           Cancel
        </button>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-secondary text-secondary-foreground shadow-sm hover:bg-secondary/80 h-9 px-4 py-2"
        >
          <svg
            class="lucide lucide-square-pen mr-2 h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
            />
            <path
              d="M18.375 2.625a2.121 2.121 0 1 1 3 3L12 15l-4 1 1-4Z"
            />
          </svg>
           Edit
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`BookingCard GIVEN a BookingCard
      WHEN render
      THEN should render all props the right way 2`] = `
<body>
  <div>
    <div
      class="rounded-xl border bg-card text-card-foreground shadow w-full"
    >
      <div
        data-radix-aspect-ratio-wrapper=""
        style="position: relative; width: 100%; padding-bottom: 56.25%;"
      >
        <div
          class="overflow-hidden mb-4 rounded-t-md"
          style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px;"
        >
          <img
            alt="Photo of room name 1"
            class="rounded-md object-cover w-full h-full"
            loading="lazy"
            src=""
          />
        </div>
      </div>
      <div
        class="space-y-1.5 p-6 flex flex-col pb-3 pt-0"
      >
        <h3
          class="font-semibold leading-none tracking-tight"
        >
          room name 1
        </h3>
        <p
          class="text-sm text-muted-foreground flex flex-row justify-between gap-x-6 flex-wrap"
        >
          <span>
            Jan 20, 2024 - Jan 24, 2024
          </span>
          <span>
            4 days
          </span>
        </p>
      </div>
      <div
        class="p-6 pt-0 flex flex-col gap-1 pb-4"
      >
        <div
          class="flex gap-6 justify-between"
        >
          <span
            class="underline"
          >
            Total
          </span>
          <span>
            $
            5250
          </span>
        </div>
        <div
          class="flex gap-6 justify-between"
        >
          <span
            class="underline"
          >
            Guests
          </span>
          <span>
            5
          </span>
        </div>
      </div>
      <div
        class="items-center p-6 pt-0 flex justify-between gap-4 flex-wrap"
      >
        <button
          aria-controls="radix-:r3:"
          aria-expanded="false"
          aria-haspopup="dialog"
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-destructive text-destructive-foreground shadow-sm hover:bg-destructive/90 h-9 px-4 py-2"
          data-state="closed"
          type="button"
        >
          <svg
            class="lucide lucide-trash mr-2 h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 6h18"
            />
            <path
              d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
            />
            <path
              d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
            />
          </svg>
           Cancel
        </button>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-secondary text-secondary-foreground shadow-sm hover:bg-secondary/80 h-9 px-4 py-2"
        >
          <svg
            class="lucide lucide-square-pen mr-2 h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
            />
            <path
              d="M18.375 2.625a2.121 2.121 0 1 1 3 3L12 15l-4 1 1-4Z"
            />
          </svg>
           Edit
        </button>
      </div>
    </div>
  </div>
</body>
`;
